File: number_of_provinces.py
#https://leetcode.com/problems/number-of-provinces
#dfs #graph #recurssion
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        reachedState = [False] * len(isConnected)
        count = 0
        for idx in range(len(isConnected)):
            if not reachedState[idx]:
                count += 1
                reachedState[idx] = True
                self.expend(isConnected, reachedState, idx)
        return count

    def expend(self, 
               isConnected: List[List[int]],
               reachedState: List[bool],
               rowIdx: int):
        for colIdx in range(len(isConnected[rowIdx])):
            if isConnected[rowIdx][colIdx]:
                if not reachedState[colIdx]:
                    reachedState[colIdx] = True
                    self.expend(isConnected, reachedState, colIdx)

File: in_progress/number_of_provinces.cc
/**
 * https://leetcode.com/problems/number-of-provinces
 * #graph #dfs
 */

class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int count = 0;
        vector<bool> reachedState(isConnected.size(), false);
        for(int idx = 0; idx < reachedState.size(); idx++){
            if(!reachedState[idx]){
                count++;
                reachedState[idx] = true;
                expand(isConnected, reachedState, idx);
            }
        }
        return count;
    }

    void expand(const vector<vector<int>>& isConnected,
           vector<bool>& reachedState,
           int rowIdx){
        for(int colIdx = 0; colIdx < isConnected[rowIdx].size(); colIdx++){
            if(isConnected[rowIdx][colIdx] && !reachedState[colIdx]){
                reachedState[colIdx] = true;
                expand(isConnected, reachedState, colIdx);
            }
        }
    }
};
